### PR TITLE
Allow extending/customizing Services

### LINKED ISSUE #346: Allow extending/customizing Services
##### **Overview of the feature request**

The idea is to have one option for customizing/extending service classes (a common use case). 
When using the generator to update existing entity or upgrading jhipster version it overwrites any customization added to the generated service files.

It would be nice to be able to customize/extend service implementations and not worry about having custom code overwritten when using the generator.

This can be done using reflection/assembly scanning to register custom implementations with the DI container.

I have used Scrutor library (https://github.com/khellang/Scrutor) for assembly scanning, it uses reflection under the hood, works with the default .Net core DI container, simplifying assembly scanning and service registration. This could also be done with using Reflection directly but the code wolud be much more complex. No need to reinvent the wheel if already exists a library that does this well.

The ServiceStartup code I used, for reference. (might need some changes):
```
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.DependencyModel;
using MyCompany.Domain.Services;
using MyCompany.Domain.Services.Interfaces;
using Scrutor;

namespace MyCompany.Configuration {
    public static class ServiceStartup
    {
        public static IServiceCollection AddServiceModule(this IServiceCollection @this)
        {
            // services will be added here by the generator
            @this.Scan(scan => scan
                // find services and register matching interfaces (EmployeeServices matches IEmployeeService)
                .FromAssembliesOf(typeof(AssemblyOfServiceInterfacesHelperType), typeof(AssemblyOfServiceClassesHelperType))
                    .AddClasses(classes => classes.InNamespaces("MyCompany.Domain.Services"))
                        .UsingRegistrationStrategy(RegistrationStrategy.Replace(ReplacementBehavior.ServiceType))
                        .AsMatchingInterface()
                        .WithScopedLifetime()

                    // now find services with name ending with 'ExtendedService' and register any interfaces it implements
                    // replacing default services (register JobExtendedService as implementation of IJobService)
                    .AddClasses(classes => classes.Where(type => type.Namespace.Equals("MyCompany.Domain.Services") &&
                                                                type.Name.EndsWith("ExtendedService")))
                        .UsingRegistrationStrategy(RegistrationStrategy.Replace(ReplacementBehavior.ServiceType))
                        .AsImplementedInterfaces()
                        .WithScopedLifetime()                    
            );
            return @this;
        }
    }
}
```

Summing up: 
If EmployeeExtendedService is present, this is the class that would be used for IEmployeeService interface on DI container. 
If no EmployeeExtendedService is present, use the generated one (EmployeeService).

Documentation should also be added on how to customize/extend services.

I can provide a PR when I have some time to add this to the generator.

##### **Motivation for or Use Case**

This allows to upgrade jhipster or change entity and have your custom service code untouched.
It is annoying to upgrade jhipster or edit one entity and have custom service code overwritten/replaced and need revert it.

Use case:
The developer needs to customize the service implementation, adding, for example, custom business logic to the EmployeeService.
The developer adds a new EmployeeExtendedService class, extending EmployeeService and adding custom business logic to one of the methods, thus, changing its behaviour.
On application startup, look for Service interfaces at the ProjectName.Domain.Services.Interfaces assembly/namespace and service implementations at ProjectName.Domain.Services assembly/namespace.
Then register the DI interface and implementation for them, with the EmployeeExtendedService class having pripority over the generated class.
No need to register interfaces/implementations on DI container like `@this.AddScoped<IEmployeeService, EmployeeService>();`.
Now the developer can confidently change the entity and/or upgrade jhipster version without losing customizations to service, since his custom service implementation will not be overwritten. And no need to manually register it.

##### **Related issues or PR**

-   [x] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by ivanmonteiro**
Fix #346

-   Please make sure the below checklist is followed for Pull Requests.

-   [x] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
-   [x] Tests are added where necessary
-   [x] Documentation is added/updated where necessary
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

**Review by nicolas63 — COMMENTED** [review]
Don't forget to set methods of services virtual for overriding later

**Comment by nicolas63** [comment]
@ivanmonteiro this PR is in draft do you need help/review ?

**Comment by ivanmonteiro** [comment]
> @ivanmonteiro this PR is in draft do you need help/review ?

I still need to add documentation on how to use this, as I found myself without time to do it, but I hope to provide the documentation in a couple of days, then It will be ready for review

**Comment by ivanmonteiro** [comment]
@nicolas63 I have added documentation and tests and maked as ready for review

**Comment by nicolas63** [comment]
LGTM you can merge it 👍 . The readme is starting to get big I think it's time we started a wiki or a site like the kotlin blueprint

**Comment by nicolas63** [comment]
Hello i have see your pipeline in error after my merge request on blazor. How can i help you ? I have changed default namespace maybe you can verify this ( default before : mycompany, default after : applicationname)

File: test-integration/samples/csharp-di-test/CountryExtendedService.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
+using System.Threading.Tasks;
+using JHipsterNet.Core.Pagination;
+using JHipsterNet.Core.Pagination.Extensions;
+using MyCompany.Data.Extensions;
```
Comment (nicolas63) [inline test-integration/samples/csharp-di-test/CountryExtendedService.cs:4]: Maybe the error is due to this ?

Comment (ivanmonteiro) [reply test-integration/samples/csharp-di-test/CountryExtendedService.cs:4]: You're right, trying to fix it soon

**Comment by nicolas63** [comment]
Thanks, and sorry for my big PR

**Comment by ivanmonteiro** [comment]
> Thanks, and sorry for my big PR

No problem. It was a nice contribution :)

**Pull request merged**